src/library_syscall.js

```diff
@@ -722,16 +721,17 @@ var SyscallsLibrary = {
           }
           throw e;
         }
-        id = child.id;
-        type = FS.isChrdev(child.mode) ? 2 :  // DT_CHR, character device.
-               FS.isDir(child.mode) ? 4 :     // DT_DIR, directory.
-               FS.isLink(child.mode) ? 10 :   // DT_LNK, symbolic link.
+        type = FS.isChrdev(node.mode) ? 2 :  // DT_CHR, character device.
+               FS.isDir(node.mode) ? 4 :     // DT_DIR, directory.
+               FS.isLink(node.mode) ? 10 :   // DT_LNK, symbolic link.
                8;                             // DT_REG, regular file.
       }
-#if ASSERTIONS
-      assert(id);
-#endif
-      {{{ makeSetValue('dirp + pos', C_STRUCTS.dirent.d_ino, 'id', 'i64') }}};
+      // We use getattr to decide which ino we return to stat, so make sure to
+      // use it to decide which ino we return to readdir too. noderawfs puts the
+      // native inode into node.id and doesn't define node_ops.getattr so use
+      // that as a fallback.
+      var ino = node.node_ops.getattr?.(node).ino ?? node.id;
```

Comment:
This `getattr` will fail if the directory doesn't have execute permissions. So perhaps the better approach is to return the virtual file system inode from stat instead of the native inode.